File: benchmarks/programs/python/strings_build.py
#!/usr/bin/env python3
# Benchmark: idiomatic string building via list-append + join (O(n)).
# n=1_000_000 — identical across Skarn, Python, C++.
# Measures the builder path; strings.py measures the O(n^2) concat path.
import time, gc

_gc_ns = 0; _gc_start = 0
def _gc_cb(phase, info):
    global _gc_ns, _gc_start
    if phase == 'start': _gc_start = time.perf_counter_ns()
    else: _gc_ns += time.perf_counter_ns() - _gc_start
gc.callbacks.append(_gc_cb)

def main():
    gc_before = sum(s['collections'] for s in gc.get_stats())
    n = 1_000_000
    t0 = time.perf_counter_ns()
    parts = []
    for _ in range(n):
        parts.append("x")
    s = "".join(parts)
    wall_ns = time.perf_counter_ns() - t0
    gc_collections = sum(s2['collections'] for s2 in gc.get_stats()) - gc_before
    print(f"benchmark=strings_build  n={n}  len={len(s)}  wall_ns={wall_ns}  ns/iter={wall_ns // n}")
    print(f"gc_collections={gc_collections}  gc_bytes=0  gc_ns={_gc_ns}")

if __name__ == "__main__":
    main()

File: benchmarks/programs/cpp/strings_build.cpp
// Benchmark: idiomatic string building via operator+= on std::string (O(n) amortized).
// n=1_000_000 — identical across Skarn, Python, C++.
// Measures the builder path; strings.cpp measures the O(n^2) concat path.
#include <cstdio>
#include <chrono>
#include <string>

int main() {
    static volatile long long _n_runtime = 1'000'000LL;
    const long long n = _n_runtime;

    auto t0 = std::chrono::high_resolution_clock::now();

    std::string s;
    for (long long i = 0; i < n; ++i) s += 'x';

    auto wall_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::high_resolution_clock::now() - t0).count();

    std::printf("benchmark=strings_build  n=%lld  len=%zu  wall_ns=%lld  ns/iter=%lld\n",
                n, s.size(), wall_ns, wall_ns / n);
    std::printf("gc_collections=0  gc_bytes=0  gc_ns=0\n");
    return 0;
}
